### ISSUE TITLE
Select For Update blocks follower reads from the leaseholder node.

### ISSUE THREAD

**Issue by data-matt** [cockroachdb/cockroach#77476]
**Describe the problem**

Select For Update blocks follower reads from the node that is currently hosting the leaseholder.

Say we have a 3 node cluster. 

We want to use SELECT FOR UPDATE for 1 transaction, whilst in a parallel transactions run follower read based queries. 

Leaseholder for the range is on Node1.

1. Connect to any node and execute: tx1 `select * from test where c2='1' for update;`
2. Connect to Node1 and start a  parallel transaction tx2 `BEGIN AS OF SYSTEM TIME follower_read_timestamp();`
3. tx2 will hang

On the flip side:
1. Connect to any node and execute: tx1 `select * from test where c2='1' for update;`
2. Connect to Node2/3 and start a  parallel transaction tx2 `BEGIN AS OF SYSTEM TIME follower_read_timestamp();`
3. tx2 will complete.

**To Reproduce**
Start a 3 node cluster.

**Terminal 1**
```
create table test ( c1 uuid, c2 string);
insert into test select gen_random_uuid(), cast(generate_series as string) from generate_series(1,100000);
begin;
select * from test where c2='1' for update;
```

**Terminal 2**
```
BEGIN AS OF SYSTEM TIME follower_read_timestamp();
select * from test where c2='1';
```

**Expected behavior**
tx1 completes from any node, as it is using follower reads.
Perhaps in the event the of SFU, divert follower reads to other nodes? 

**Environment:**
 - CockroachDB version 21.1.16
 - Server OS: Ubuntu
 - Client app: cockroach sql

**Comment by rharding6373** [comment]
Could someone from the KV team please verify that this is expected behavior? Thanks in advance!

**Comment by nvanbenschoten** [comment]
This is semi-expected behavior, though it does lead to the surprising behavior mentioned above. SELECT FOR UPDATE acquires unreplicated locks on the leaseholder only, by design. As mentioned in other issues (like https://github.com/cockroachdb/cockroach/issues/49684), these locks currently block other locking and non-locking reads. There are tradeoffs to that decision, which are discussed in #49684. If SELECT FOR UPDATE locks did not block non-locking reads then we could get more concurrency in some cases (like this one), but would also introduce more transaction retries under contention, which SFU is meant to limit.

The surprising thing we see with this issue is an asymmetry between when SFU locks block non-locking reads. They do on the leaseholder even for reads below the closed timestamp but don't on followers below the closed timestamp (simply because the locks aren't replicated). There's certainly less of a benefit for SFU locks to block non-locking reads below the closed timestamp, because any write by the locking txn that replaces the SFU lock will need to advance its write timestamp, regardless of the lock.

So I think we could change the policy to only block non-locking reads on SFU locks that are above the closed timestamp. But the layering of these concerns in the system doesn't make this easy to determine when a reader begins to block, nor does it make it easy to react to the gradual advance of the closed timestamp, which I guess should eventually permit a reader to stop blocking.

@data-matt could you describe the user impact of this issue?

**Comment by data-matt** [comment]
@nvanbenschoten I don't believe that the user group that encountered the issue is blocked by this right now. They are most likely able to workaround this by avoiding the leaseholder.

Perhaps if one of the application teams run into it, it will be a different story.  However for end users that connect via loadbalancer, it will be impossible for them to consistently workaround it, as they won't be able to specify which node they communicate with. 

I will seek an update from the customer directly.

**Comment by data-matt** [comment]
@williamkulju / @nvanbenschoten, the staleness github-action marked this inactive. I removed the stale label and this remains a surprise and a bug.
